### ISSUE TITLE
GCD of Gaussian integers

### ISSUE THREAD

**Issue by normalhuman** [sympy/sympy#15358]
```
>>> gcd(5, 2+I, gaussian=True)
1
```
But `2+I` would be correct here, since `5 = (2+I)*(2-I)`. 

Aside: the gaussian flag works for monic polynomials as expected. 
```
>>> var('x')
>>> gcd(x**2 + 9, x + 3*I)
1
>>> gcd(x**2 + 9, x + 3*I, gaussian=True)
x + 3*I
```

From [Stack Overflow](https://stackoverflow.com/questions/52692068/find-the-greatest-common-divisor-gcd-in-gaussian-integers-with-sympy).

**Comment by jksuom** [comment]
`gaussian=True` has been designed as an alias of `domain=QQ(I)`, which is a field. Hence 1 could be the expected result of `gcd(5, 2+I, gaussian=True)`. Another keyword would probably be needed for those finite extensions of `ZZ` where `gcd` can be implemented.

**Comment by asmeurer** [comment]
`gcd` already special-cases constant polynomials, performing integer gcd (or even generalizations for non-integers). Is there any reason to not do the same for gaussian integers/rationals?

**Comment by jksuom** [comment]
That could probably be done (though I think that special-casing integers is confusing). The class of Gaussian integers/rationals should be constructed, [this comment](https://github.com/sympy/sympy/blob/master/sympy/polys/constructor.py#L22) should be effected and gcd implemented.

**Comment by normalhuman** [comment]
Understood as working over QQ(I), one may expect 1, but then there is also this: 
```
>>> gcd(4+2*I, 2*I, gaussian=True)
2
```

**Comment by czgdp1807** [comment]
The issue still persists on `master`.

**Comment by oscarbenjamin** [comment]
This issue is fixed now:
```python
In [1]: gcd(5, 2+I, gaussian=True)
Out[1]: 2 + ⅈ
```
The fix was 209fd309627836ee8936eb420279aeb4f341e960 from gh-19485.

**Comment by oscarbenjamin** [comment]
> Fastest Gaussian Integer GCD algorithms from the literature:

Feel free to open a new issue discussing a dfferent algorithm for Gaussian integer gcd. Currently Euclid's algorithm is used:
https://github.com/sympy/sympy/blob/fde616dae12f576885711d736039afe776540488/sympy/polys/domains/gaussiandomains.py#L459-L463

**Comment by Hermann-SW** [comment]
Correction, above script measures more than gaussian gcd().

The new gcd() benchmark compares Rober Chapman ggcd() and sympy gaussian integer gcd(), with and without use of gmpy2, on Intel and Arm Cortex-A72:  
https://github.com/Hermann-SW/RSA_numbers_factored/blob/main/python/10000.gaussian_gcd.py

sympy gaussian integer gcd() gets minimally slower with mpz number instead of Python int, while ggcd() gets much faster witth gmpy2, even (a little) faster than sympy gcd(), on both CPU architectures.

```
$ grep i7 /proc/cpuinfo | head -1 | cut -f2 -d:
 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
$ uname -sr
Linux 4.18.0-425.19.2.el8_7.x86_64
$ python --version
Python 3.9.13
$ 
$ python 10000.gaussian_gcd.py 
ggcd((L, 0), (sqrtm1, 1)): 3.773860454559326s
gcd(L, sqrtm1 + I): 0.7253265380859375s
$ 
$ python 10000.gaussian_gcd.py gmpy2
gmpy2
ggcd((L, 0), (sqrtm1, 1)): 0.6698856353759766s
gcd(L, sqrtm1 + I): 0.7326250076293945s
$ 
```

1.8GHz Arm Cortex-A72:  
```
pi@pi400-64:~/RSA_numbers_factored/python $ uname -sr
Linux 6.1.21-v8+
pi@pi400-64:~/RSA_numbers_factored/python $ python --version
Python 3.9.2
pi@pi400-64:~/RSA_numbers_factored/python $ python 10000.gaussian_gcd.py 
ggcd((L, 0), (sqrtm1, 1)): 10.806464672088623s
gcd(L, sqrtm1 + I): 3.194443702697754s
pi@pi400-64:~/RSA_numbers_factored/python $ 
pi@pi400-64:~/RSA_numbers_factored/python $ python 10000.gaussian_gcd.py gmpy2
gmpy2
ggcd((L, 0), (sqrtm1, 1)): 2.995866298675537s
gcd(L, sqrtm1 + I): 3.1801211833953857s
pi@pi400-64:~/RSA_numbers_factored/python $ 
```

**Comment by Hermann-SW** [comment]
No, the "pow()" time for determining ```sqrtm1``` and asserting identities is not measured.
Only calls to ```gcd()``` and ```ggcd()``` are measured, so this is a pure greatest common divisor benchmark:
```
....
start = time()
t = ggcd((L, 0), (sqrtm1, 1))
mid = time()
t2 = gcd(L, sqrtm1 + I)
end = time()
....
```

**Comment by Hermann-SW** [comment]
I did benchmark sympy gaussian gcd() on "really big" numbers (real parts up to 388342-digit numbers).
Again ```sqrt(-1) (mod p)``` was precomputed, so only ```gcd()``` runtime got measured.
The runtime is quadratic in the input size (as [quadratic regression](https://github.com/Bumpkin-Pi/QuadraticRegression/tree/master) shows), with 42min runtime for 388342-digit prime:  
```
$ python main.py gcd.csv 
['main.py', 'gcd.csv']
y = 1.7552402091077303e-08x² - 0.00031564680429543857x + 0.3202998202541422
$
```
<img src="https://stamm-wilbrandt.de/images/gaussian_integer_gcd.benchmark.png">

**Comment by Hermann-SW** [comment]
I added two more data points between 200700-digit and 388342-digit primes, found here:  
https://t5k.org/primes/lists/all.txt
```
...
-----  ------------------------------- -------- ----- ---- --------------
 rank  description                     digits   who   year comment
-----  ------------------------------- -------- ----- ---- --------------
...
 5058  2145*2^1099064+1                  330855 L1792 2013 Divides Fermat F(1099061)
....
 5068  1705*2^906110+1                   272770 L3174 2012 Divides Fermat F(906108)
...
```
More details in [Sympy gaussian integer gcd() benchmark](https://github.com/Hermann-SW/RSA_numbers_factored/tree/main/python/sympy_gaussian_integer_gcd_benchmark#sympy-gaussian-integer-gcd-benchmark):
<img src="https://github.com/Hermann-SW/RSA_numbers_factored/raw/main/python/sympy_gaussian_integer_gcd_benchmark/sgig_benchmark.png">
